### PR TITLE
Support certificate policies in the pki backend

### PR DESCRIPTION
**Pull request by remip2**
Add policy_identifiers field to the pki role to generate certificate with custom certificate policies.
Add basic constraint CA:FALSE by default in certificate extensions.

### CODE REVIEW

File: builtin/logical/pki/cert_util.go [hunk 1]
```diff
@@ -1240,6 +1259,9 @@ func signCertificate(data *dataBundle) (*certutil.ParsedCertBundle, error) {
 		if certTemplate.MaxPathLen == 0 {
 			certTemplate.MaxPathLenZero = true
 		}
+	} else {
+		certTemplate.BasicConstraintsValid = true
```
Comment (jefferai) [inline builtin/logical/pki/cert_util.go]: What's the reason for this change? Also, IsCA wouldn't need to be set explicitly false, and if BasicConstraintsValid should be true in either case it should be pulled out of the if condition.

Comment (remip2) [reply builtin/logical/pki/cert_util.go]: client/server certificates should have this constraint explicitly set. Some old tools/browser refuse to connect to web servers with a certificate without the basicConstraint CA:False

To not change the default behaviour, I updated the patch to add an option in the pki role for that.

File: builtin/logical/pki/cert_util.go [hunk 2]
```diff
@@ -985,6 +989,19 @@ func addKeyUsages(data *dataBundle, certTemplate *x509.Certificate) {
 	}
 }
 
+// addPolicyIdentifiers adds certificate policies extension
+//
+func addPolicyIdentifiers(data *dataBundle, certTemplate *x509.Certificate) {
+	if len(data.params.PolicyIdentifiers) > 0 {
```
Comment (vishalnayak) [inline builtin/logical/pki/cert_util.go]: Can this `if` condition can be removed?

File: builtin/logical/pki/cert_util.go [hunk 3]
```diff
@@ -985,6 +989,17 @@ func addKeyUsages(data *dataBundle, certTemplate *x509.Certificate) {
 	}
 }
 
+// addPolicyIdentifiers adds certificate policies extension
+//
+func addPolicyIdentifiers(data *dataBundle, certTemplate *x509.Certificate) {
+	for _, oidstr := range data.params.PolicyIdentifiers {
+		oid, err := stringToOid(oidstr)
+		if err == nil {
```
Comment (chrishoffman) [inline builtin/logical/pki/cert_util.go]: I think we should return an error if we can't parse an OID instead of silently swallowing the error.

File: builtin/logical/pki/path_roles.go [hunk 4]
```diff
@@ -262,6 +262,16 @@ for "generate_lease".`,
 				Default:     true,
 				Description: `If set to false, makes the 'common_name' field optional while generating a certificate.`,
 			},
+			"policy_identifiers": &framework.FieldSchema{
+				Type:        framework.TypeCommaStringSlice,
+				Default:     []string{},
```
Comment (chrishoffman) [inline builtin/logical/pki/path_roles.go]: This default value is not required.

File: builtin/logical/pki/path_roles.go [hunk 5]
```diff
@@ -262,6 +262,16 @@ for "generate_lease".`,
...
+				Description: `A comma-separated string or list of policy oids.`,
+			},
+			"basic_constraints_valid_for_non_ca": &framework.FieldSchema{
+				Type:        framework.TypeBool,
+				Default:     false,
```
Comment (chrishoffman) [inline builtin/logical/pki/path_roles.go]: This default value is not required.

File: builtin/logical/pki/cert_util.go (hunk 3 above)
Comment (jefferai) [reply builtin/logical/pki/cert_util.go]: Do you think we should do that at submission time, sanity-checking the role data that was provided? It's safer to do in both places but not sure it's really necessary.

**Comment by jefferai** [comment]
Going to merge this, will address Chris' concern separately.

**Comment by jefferai** [comment]
Thanks @remip2 !
